app/controllers/users_controller.rb

```diff
@@ -24,16 +24,41 @@ def edit
   end
 
   def update
-    if @user.update(user_params)
-      redirect_to edit_user_path(@user), notice: "You have successfully updated your user data."
-    elsif user_params[:password] === ''
-      flash.now[:error] = "Password is a mandatory field"
+    if user_params[:password] === ''
```

Comment:
I would also create a "validate_password" method here that would look like this:
```
  def validate_password
    if user_params[:password] === ''
      flash[:error] = "Password is a mandatory field"
      redirect_to edit_user_path(@user)
    elsif @user.authenticated?(params[:user][:password]) == false
      flash[:error] = "Incorrect password"
      redirect_to edit_user_path(@user)
    end
  end
```
And skip the whole password validation in the update action. This way we'll have a cleaner update code and will only be in charge of updating data, meanwhile the validate_password will be in charge of checking the old password field. The update action would then look like this:
```
def update
    validate_password
    if @user.update(user_params) && params[:commit] == "Delete Account"
      redirect_to delete_account_path(@user)
    elsif @user.update(user_params)
      if user_params[:new_password] != ''
        @user.update_attributes(password: user_params[:new_password])
      end
      redirect_to edit_user_path(@user), notice: "You have successfully updated your user data."
    end
  end
```

And we should add validate_password to the before action ensure_correct_user on the top of the user controller.

Reply:
(For the same reason as in the other PR that I commented before ❤️ )